Cartridge: v11 JIT fails during install for adapter.cpp/adapter.bc

**Description:**

- RDKit Version: 201803.4
- Platform: Debian unstable

Buliding against PostgreSQL 11 using the PGXS Makefile works fine, but I get an error during install:

```
/bin/mkdir -p '/<<BUILDDIR>>/rdkit-201803.4+dfsg/debian/postgresql-11-rdkit/usr/lib/postgresql/11/lib/bitcode/rdkit'
/bin/mkdir -p '/<<BUILDDIR>>/rdkit-201803.4+dfsg/debian/postgresql-11-rdkit/usr/lib/postgresql/11/lib/bitcode'/rdkit/
/usr/bin/install -c -m 644 rdkit_io.bc '/<<BUILDDIR>>/rdkit-201803.4+dfsg/debian/postgresql-11-rdkit/usr/lib/postgresql/11/lib/bitcode'/rdkit/./
/usr/bin/install -c -m 644 mol_op.bc '/<<BUILDDIR>>/rdkit-201803.4+dfsg/debian/postgresql-11-rdkit/usr/lib/postgresql/11/lib/bitcode'/rdkit/./
/usr/bin/install -c -m 644 bfp_op.bc '/<<BUILDDIR>>/rdkit-201803.4+dfsg/debian/postgresql-11-rdkit/usr/lib/postgresql/11/lib/bitcode'/rdkit/./
/usr/bin/install -c -m 644 sfp_op.bc '/<<BUILDDIR>>/rdkit-201803.4+dfsg/debian/postgresql-11-rdkit/usr/lib/postgresql/11/lib/bitcode'/rdkit/./
/usr/bin/install -c -m 644 rxn_op.bc '/<<BUILDDIR>>/rdkit-201803.4+dfsg/debian/postgresql-11-rdkit/usr/lib/postgresql/11/lib/bitcode'/rdkit/./
/usr/bin/install -c -m 644 rdkit_gist.bc '/<<BUILDDIR>>/rdkit-201803.4+dfsg/debian/postgresql-11-rdkit/usr/lib/postgresql/11/lib/bitcode'/rdkit/./
/usr/bin/install -c -m 644 bfp_gist.bc '/<<BUILDDIR>>/rdkit-201803.4+dfsg/debian/postgresql-11-rdkit/usr/lib/postgresql/11/lib/bitcode'/rdkit/./
/usr/bin/install -c -m 644 bfp_gin.bc '/<<BUILDDIR>>/rdkit-201803.4+dfsg/debian/postgresql-11-rdkit/usr/lib/postgresql/11/lib/bitcode'/rdkit/./
/usr/bin/install -c -m 644 low_gist.bc '/<<BUILDDIR>>/rdkit-201803.4+dfsg/debian/postgresql-11-rdkit/usr/lib/postgresql/11/lib/bitcode'/rdkit/./
/usr/bin/install -c -m 644 guc.bc '/<<BUILDDIR>>/rdkit-201803.4+dfsg/debian/postgresql-11-rdkit/usr/lib/postgresql/11/lib/bitcode'/rdkit/./
/usr/bin/install -c -m 644 cache.bc '/<<BUILDDIR>>/rdkit-201803.4+dfsg/debian/postgresql-11-rdkit/usr/lib/postgresql/11/lib/bitcode'/rdkit/./
/usr/bin/install -c -m 644 adapter.bc '/<<BUILDDIR>>/rdkit-201803.4+dfsg/debian/postgresql-11-rdkit/usr/lib/postgresql/11/lib/bitcode'/rdkit/./
/usr/bin/install -c -m 644 bitstring.bc '/<<BUILDDIR>>/rdkit-201803.4+dfsg/debian/postgresql-11-rdkit/usr/lib/postgresql/11/lib/bitcode'/rdkit/./
cd '/<<BUILDDIR>>/rdkit-201803.4+dfsg/debian/postgresql-11-rdkit/usr/lib/postgresql/11/lib/bitcode' && /usr/lib/llvm-7/bin/llvm-lto -thinlto -thinlto-action=thinlink -o rdkit.index.bc rdkit/rdkit_io.bc rdkit/mol_op.bc rdkit/bfp_op.bc rdkit/sfp_op.bc rdkit/rxn_op.bc rdkit/rdkit_gist.bc rdkit/bfp_gist.bc rdkit/bfp_gin.bc rdkit/low_gist.bc rdkit/guc.bc rdkit/cache.bc rdkit/adapter.bc rdkit/bitstring.bc
error: can't create module summary index for buffer: Expected a single module
LLVM ERROR: ThinLink didn't create an index
make[2]: *** [/usr/lib/postgresql/11/lib/pgxs/src/makefiles/pgxs.mk:219: install] Error 1
make[2]: Leaving directory '/<<BUILDDIR>>/rdkit-201803.4+dfsg/build-11'
make[1]: *** [debian/rules:94: override_dh_install] Error 2
```

If I remove `adapter.bc` (the sole C++ file) from the `llvm-lto` command-line, the command works fine. If I run `llvm-dis-7` on `adapter.bc`, I get:

```
llvm-dis-7: error: Expected a single module
```

Work-around for now is to add `override with_llvm=no` in `Code/PgSQL/rdkit/Makefile`, but it would be nice if JIT could be enabled for the Cartridge.

Not quite sure what the error means, but maybe you have an idea?

--- a/Code/PgSQL/rdkit/Makefile
+++ b/Code/PgSQL/rdkit/Makefile
@@ -15,7 +15,7 @@ RDKIT=$(RDBASE)
 
 
 ifeq ($(USE_INCHI),1)
-  INCHILIBS=-lRDInchiLib -lInchi
+  INCHILIBS=-lRDKitRDInchiLib -lRDKitInchi
   INCHIFLAGS=-I${RDKIT}/External -DBUILD_INCHI_SUPPORT
   INCHIREGRESS=inchi
   USE_THREADS=1
@@ -32,7 +32,7 @@ ifeq ($(USE_THREADS),1)
   THREADLIBS=-L${BOOSTHOME}/lib -lboost_thread -lboost_system
 endif
 
-RDKLIBS       = ${AVALONLIBS} ${INCHILIBS} -lMolDraw2D -lMolTransforms -lMolHash -lFMCS -lChemReactions -lChemTransforms -lFileParsers -lSmilesParse -lFingerprints -lSubgraphs -lDescriptors -lPartialCharges -lSubstructMatch  -lGraphMol -lEigenSolvers -lDataStructs -lDepictor -lRDGeometryLib -lRDGeneral
+RDKLIBS       = ${AVALONLIBS} ${INCHILIBS} -lRDKitMolDraw2D -lRDKitMolTransforms -lRDKitMolHash -lRDKitFMCS -lRDKitChemReactions -lRDKitChemTransforms -lRDKitFileParsers -lRDKitSmilesParse -lRDKitFingerprints -lRDKitSubgraphs -lRDKitDescriptors -lRDKitPartialCharges -lRDKitSubstructMatch  -lRDKitGraphMol -lRDKitEigenSolvers -lRDKitDataStructs -lRDKitDepictor -lRDKitRDGeometryLib -lRDKitRDGeneral
 
 ifeq ($(STATIC_LINK),0)
   SHLIB_LINK += -L${RDKIT}/lib -Wl,-rpath,'${RDKIT}/lib' ${RDKLIBS}
@@ -101,3 +101,10 @@ CC = $(CXX)
 
 %.o : %.cpp
 	$(CXX) $(CPLUSPLUSFLAGS) $(CPPFLAGS) -fPIC -c -o $@ $<
+
+
+COMPILE.cxx.bc = $(CLANG) -xc++ -Wno-ignored-attributes $(BITCODE_CXXFLAGS) $(CPPFLAGS) -emit-llvm -c
+
+%.bc : %.cpp
+	$(COMPILE.cxx.bc) -o $@ $<
+	$(LLVM_BINPATH)/opt -module-summary -f $@ -o $@

--- a/Code/PgSQL/rdkit/README
+++ b/Code/PgSQL/rdkit/README
@@ -132,4 +132,4 @@ Example:
 
 Further Reading:
 
-Cartridge tutorial [Docs/Book/Cartridge.md](../../../Docs/Book/Cartridge.md)
+Cartridge tutorial (https://github.com/rdkit/rdkit/blob/master/Docs/Book/Cartridge.md)

--- a/Code/PgSQL/rdkit/README
+++ b/Code/PgSQL/rdkit/README
@@ -8,6 +8,10 @@ Compatibility: PostgreSQL 9.1+
  that you have the -devel package installed.
 
 Installation:
+Cartirdge is build from the root RDKit level ie. you need to tell CMake to 
+build it too by including the following switch:
+    -D RDK_BUILD_PGSQL=ON
+
 if PostgreSQL is installed in a location where CMake is unable to find it,
 either set the environment variable PostgreSQL_ROOT or add a
 "-DPostgreSQL_ROOT" switch to the cmake command line, pointing to the
@@ -32,7 +36,7 @@ After building the RDKit, carry out the following steps:
    -w stop (Windows)
    service postgresql stop (Linux)
    
-3) run the "pgsql_install" script:
+3) run the "pgsql_install" script (note that it is built by the root CMake):
    
    ${CMAKE_BINARY_DIR}\Code\PgSQL\rdkit\pgsql_install.bat (Windows)
    sh ${CMAKE_BINARY_DIR}/Code/PgSQL/rdkit/pgsql_install.sh (Linux)
@@ -125,3 +129,7 @@ Indexes:
 Example:
 
   See sql/* for examples.
+
+Further Reading:
+
+Cartridge tutorial [Docs/Book/Cartridge.md](../../../Docs/Book/Cartridge.md)

--- a/Code/PgSQL/rdkit/README
+++ b/Code/PgSQL/rdkit/README
@@ -8,7 +8,7 @@ Compatibility: PostgreSQL 9.1+
  that you have the -devel package installed.
 
 Installation:
-Cartirdge is build from the root RDKit level ie. you need to tell CMake to 
+Cartridge is build from the root RDKit level ie. you need to tell CMake to 
 build it too by including the following switch:
     -D RDK_BUILD_PGSQL=ON
 

--- a/Code/PgSQL/rdkit/README
+++ b/Code/PgSQL/rdkit/README
@@ -8,7 +8,7 @@ Compatibility: PostgreSQL 9.1+
  that you have the -devel package installed.
 
 Installation:
-Cartridge is build from the root RDKit level ie. you need to tell CMake to 
+Cartridge is build from the root RDKit level, i.e. you need to tell CMake to 
 build it too by including the following switch:
     -D RDK_BUILD_PGSQL=ON
 

